src/UrlGenerator.php

```diff
@@ -0,0 +1,224 @@
+
+namespace Yiisoft\Router\FastRoute;
+
+use Nyholm\Psr7\Uri;
+use Yiisoft\Router\RouteCollectionInterface;
+use Yiisoft\Router\RouteNotFoundException;
+use Yiisoft\Router\UrlMatcherInterface;
+use Yiisoft\Router\UrlGeneratorInterface;
+use FastRoute\RouteParser;
+
+use function array_key_exists;
+use function array_keys;
+use function implode;
+use function is_string;
+use function preg_match;
+
+final class UrlGenerator implements UrlGeneratorInterface
+{
+    private string $uriPrefix = '';
+    private RouteCollectionInterface $routeCollection;
+    private UrlMatcherInterface $matcher;
+    private RouteParser $routeParser;
+
+    public function __construct(
+        UrlMatcherInterface $matcher,
+        RouteParser $parser = null
+    ) {
+        $this->matcher = $matcher;
+        $this->routeCollection = $matcher->getRouteCollection();
+        $this->routeParser = $parser ?? new RouteParser\Std();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Replacements in FastRoute are written as `{name}` or `{name:<pattern>}`;
+     * this method uses `FastRoute\RouteParser\Std` to search for the best route
+     * match based on the available substitutions and generates a uri.
+     *
+     * @throws \RuntimeException if parameter value does not match its regex.
+     */
+    public function generate(string $name, array $parameters = []): string
+    {
+        $route = $this->routeCollection->getRoute($name);
+
+        $parsedRoutes = array_reverse($this->routeParser->parse($route->getPattern()));
+        if ($parsedRoutes === []) {
+            throw new RouteNotFoundException($name);
+        }
+
+        $missingParameters = [];
+
+        // One route pattern can correspond to multiple routes if it has optional parts
+        foreach ($parsedRoutes as $parsedRouteParts) {
+            // Check if all parameters can be substituted
+            $missingParameters = $this->missingParameters($parsedRouteParts, $parameters);
+
+            // If not all parameters can be substituted, try the next route
+            if (!empty($missingParameters)) {
+                continue;
+            }
+
+            return $this->generatePath($parameters, $parsedRouteParts);
+        }
+
+        // No valid route was found: list minimal required parameters
+        throw new \RuntimeException(sprintf(
+            'Route `%s` expects at least parameter values for [%s], but received [%s]',
+            $name,
+            implode(',', $missingParameters),
+            implode(',', array_keys($parameters))
+        ));
+    }
+
+    public function generateAbsolute(string $name, array $parameters = [], string $scheme = null, string $host = null): string
+    {
+        $url = $this->generate($name, $parameters);
+        $route = $this->routeCollection->getRoute($name);
+        /** @var Uri $uri */
+        $uri = $this->matcher->getLastMatchedRequest() !== null ? $this->matcher->getLastMatchedRequest()->getUri() : null;
```

Comment:
`$matcher` is type-hinted as `UrlMatcherInterface` but `getLastMatchedRequest()` is not part of the interface. Either we should type-hint against `UrlMatcher` or add the method to the interface.

Reply:
It will be added to the interface.